### DISCUSSION TITLE
parametrized_dag_factory behaviour

### DISCUSSION THREAD

**Discussion by sylvainbonnot** [i2mint/meshed#63] [Enhancement]
## Non intuitive behavior

The behaviour of [parametrized_dag_factory](https://github.com/i2mint/meshed/blob/b387e728198195e73a16d6e8dc420500e6f37380/meshed/dag.py#L1752)
is not too intuitive: one gets the following

```python
from meshed import code_to_dag
from meshed.dag import parametrized_dag_factory

@code_to_dag
def dag1():
    a = criss(x,y)
    b = cross(aa, bb)
    c = apple(a, b)


dag_factory1 = parametrized_dag_factory(dag1, 'a')
assert str(Sig(dag_factory1(x=1,y=2))) == '(b)'  # but we DO NOT WANT THAT.
assert str(Sig(dag_factory1(x=1,y=2))) != '(aa, bb)'  # but we DO WANT THAT.
```

<img width="402" alt="image" src="https://github.com/i2mint/meshed/assets/1906276/b28d779a-b9f7-41dc-ab34-f8a0c2652ffc">

## Inconsistent behavior

Also, when a function in the dag has no params, the parametrized_dag is not consistent with the above:

```python
@code_to_dag
def dag2():
    a = criss()
    b = cross(aa, bb)
    c = apple(a, b)

dag_factory2 = parametrized_dag_factory(dag2, 'a')
assert str(Sig(dag_factory2())) == '(a, b)'
# <Sig (a, b)>  # to be consistent with the above, it should be Sig (b)
```

<img width="245" alt="image" src="https://github.com/i2mint/meshed/assets/1906276/d9ca1dcf-d7ee-49dc-9882-cf820a1fd198">

**Comment by thorwhalen** [comment]
Great observations and I agree on both:

* Indeed, my intent was that the first be `Sig (aa, bb)`
* Consistency is crucial. Without any clear argument or narrative supporting it, I'd put forth an hypothetical rule of thumb: _(Behavioral) consistency trumps intuitiveness_. 

This is a first draft of the function, so maybe it's an actual bug that we can resolve. Unfortunately, there's a chance it may be a bit more involved than that. 

Maybe it's a bug else where (ancestors? partial?). 

Or maybe, and that would suck, it's one of those "can't satisfy all expectations" kind of problem. Example; the non-obvious [C3 linearization](https://blog.pilosus.org/posts/2019/05/02/python-mro/) solution to getting a method-resolution-order (MRO) that would preserve the order of classes as defined by the programmer, while also preserving the monotonicity of the inheritance graph.

Either way, I'm moving this to a discussion so we can figure out what behavior we actually want (we now have two examples of what we **do not** want.

**Comment by thorwhalen** [comment]
The first problem comes from a bug (or an "unexpected behavior" -- TBD) of `__getitem__`. 

```python
computation_dag = dag1['a':]
computation_dag.dot_digraph()
```

<img width="157" alt="image" src="https://github.com/i2mint/meshed/assets/1906276/c0e5640d-52a0-46be-a8e3-509b0819a752">

We have:

```
assert str(Sig(computation_dag)) == '(a, b)'
```

which translates to `(b)` when `a` is removed (on purpose -- see if you agree) when doing the partial. 

Now, we need to go back to the definition of `__getitem__` and see if we agree with this `dag1['a':]`. 
It seems like `dag1['a':]` gets the subdag goind **down**, from `a`, picking up `b` because it needs it to compute `apple`. 
I can see the rationale, but not sure I agree. 
If we made that choice, it's probably because **that** was the behavior that was intuitive for the context we first made `__getitem__` for. 

So either:
* We revise the definition of `__getitem__`. We should bare in mind that this might be a bit disruptive, so we should do that only if we really determine that the choice we made in the current one was not a "good" one. One argument against the current behavior being a "good" one is that `dag[:a] + dag[a:] != dag`. 
* We keep the current `__getitem__` behavior, but add the ancestors to "complete" the graph we actually want.

**Reply by sylvainbonnot** [reply]
About the definition of ```dag[:a]``` being wrong: any definition of ```dag[:a]``` that uses only "ancestors" of ```a``` will fail to satisfy ```dag[:a] + dag[a:] != dag```. 
A simple dag to keep in mind is ```[0-> a ->2, 0 ->1]```, that is:
```mermaid
  graph TD;
      0-->a;
      a-->2;
      0-->1;
```


  A possible definition that would satisfy  ```dag[:a] + dag[a:] != dag``` is: 
> ```dag[:a]``` is the subdag made of all the nodes in ```a``` together with all the nodes that can be joined to a node in ```a``` by at least one **undirected** path **finishing** with an edge arriving in ```a```.  

Conversely, 

> ```dag[a:]``` is the subdag made of all the nodes in ```a``` together with all the nodes that can be joined to a node in ```a``` by at least one **undirected** path **starting** with an edge exiting ```a```.

**Reply by thorwhalen** [reply]
So, what we've decided here is that the `dag[:a] + dag[a:] != dag` requirement makes sense for [total orders](https://en.wikipedia.org/wiki/Total_order), it doesn't really have a canonical extension/generalization to [partial orders](https://en.wikipedia.org/wiki/Partially_ordered_set). 

Our current definition of `DAG.__getitem__` satisfies the equation when the DAG is linear (therefore defining a total order), and that is sufficient for us for now since we don't know if there is a better definition for our context.

**Comment by sylvainbonnot** [comment]
This issue is linked to two others:
* [roll_in_orphaned_dag](https://github.com/i2mint/meshed/discussions/54) : a subdag that is fully computed might be "erased" by simply leaving its leaves apparent
* the question of subdag subtraction: ideally one wants to base it on operations that are purely graph-theoretical. A subgraph subtraction is now done in [itools_subtraction](https://github.com/i2mint/meshed/blob/bc6dff561232b08c15310871619380218f4fc46f/meshed/itools.py#L549)
